### PR TITLE
Add function to get connecting AP information.

### PR DESCRIPTION
**Pull request by donghengqaz**

### CODE REVIEW

File: src/compiler/propagation/type_primitive_wifi.cc [hunk 1]
```diff
@@ -28,10 +28,10 @@ TYPE_PRIMITIVE_ANY(setup_ip)
 TYPE_PRIMITIVE_ANY(disconnect)
 TYPE_PRIMITIVE_ANY(disconnect_reason)
 TYPE_PRIMITIVE_ANY(get_ip)
-TYPE_PRIMITIVE_ANY(get_rssi)
 TYPE_PRIMITIVE_ANY(init_scan)
 TYPE_PRIMITIVE_ANY(start_scan)
 TYPE_PRIMITIVE_ANY(read_scan)
+TYPE_PRIMITIVE_ANY(get_ap_info)
```
Comment (kasperl) [inline src/compiler/propagation/type_primitive_wifi.cc:34]: We might as well start teaching the type propagator that these primitives have very well defined return types:

```suggestion
TYPE_PRIMITIVE_ARRAY(ap_info)
```

File: lib/net/wifi.toit [hunk 2]
```diff
@@ -64,6 +64,14 @@ class AccessPoint:
     return (List bssid.size: "$(%02x bssid[it])").join ":"
 
 interface Interface extends net.Interface:
+  /**
+  Return current access point information.
```
Comment (kasperl) [inline lib/net/wifi.toit:68]:
```suggestion
  Returns information about the access point this $Interface is currently 
    connected to.
```

File: lib/net/wifi.toit [hunk 3]
```diff
@@ -64,6 +64,14 @@ class AccessPoint:
...
+
+  Throws an exception if this network isn't currently connected to an
+    access point.
+  */
+  get_ap_info -> AccessPoint
```
Comment (kasperl) [inline lib/net/wifi.toit:73]: I think this should just be `access_point`. 

```suggestion
  access_point -> AccessPoint
```

File: lib/system/api/wifi.toit [hunk 4]
```diff
@@ -7,16 +7,16 @@ import system.api.network show NetworkService NetworkServiceClient
 interface WifiService extends NetworkService:
   static UUID  /string ::= "2436edc6-4cd8-4834-8ebc-ed883990da40"
   static MAJOR /int    ::= 0
-  static MINOR /int    ::= 8
+  static MINOR /int    ::= 9
```
Comment (kasperl) [inline lib/system/api/wifi.toit:10]: If you change the method numbering (not just add new ones), we should change the MAJOR version, but for now, all WiFi implementations are coupled with the API definition, so it doesn't matter.

File: lib/net/wifi.toit [hunk 5]
```diff
@@ -128,9 +136,17 @@ class WifiInterface_ extends SystemInterface_ implements Interface:
...
+        --ssid=ap_info[WIFI_SCAN_SSID_]
+        --bssid=ap_info[WIFI_SCAN_BSSID_]
+        --rssi=ap_info[WIFI_SCAN_RSSI_]
+        --authmode=ap_info[WIFI_SCAN_AUTHMODE_]
+        --channel=ap_info[WIFI_SCAN_CHANNEL_]
+
   signal_strength -> float:
-    rssi := (client_ as WifiServiceClient).rssi handle_
-    if not rssi: throw "wifi not connected in STA mode"
+    rssi := get_ap_info.rssi
```
Comment (kasperl) [inline lib/net/wifi.toit:149]: Maybe save the allocation of the `AccessPoint` instance?

```suggestion
    info := (client_ as WifiServiceClient).get_ap_info handle_
    rssi := info[WIFI_SCAN_RSSI_]
```

File: lib/system/api/wifi.toit [hunk 6]
```diff
@@ -7,16 +7,16 @@ import system.api.network show NetworkService NetworkServiceClient
...
   static CONNECT_INDEX /int ::= 1000
   connect config/Map? save/bool -> List
 
   static ESTABLISH_INDEX /int ::= 1001
   establish config/Map? -> List
 
-  static RSSI_INDEX /int ::= 1002
-  rssi handle/int -> int?
+  static GET_AP_INFO_INDEX /int ::= 1002
+  get_ap_info handle/int -> int?
```
Comment (kasperl) [inline lib/system/api/wifi.toit:19]:
```suggestion
  static AP_INFO_INDEX /int ::= 1002
  ap_info handle/int -> List
```

File: src/primitive.h [hunk 7]
```diff
@@ -307,10 +307,10 @@ namespace toit {
   PRIMITIVE(disconnect, 2)                   \
   PRIMITIVE(disconnect_reason, 1)            \
   PRIMITIVE(get_ip, 1)                       \
-  PRIMITIVE(get_rssi, 1)                     \
   PRIMITIVE(init_scan, 1)                    \
   PRIMITIVE(start_scan, 4)                   \
   PRIMITIVE(read_scan, 1)                    \
+  PRIMITIVE(get_ap_info, 1)                  \
```
Comment (kasperl) [inline src/primitive.h:313]:
```suggestion
  PRIMITIVE(ap_info, 1)                      \
```

File: src/resources/wifi_esp32.cc [hunk 8]
```diff
@@ -604,6 +589,35 @@ PRIMITIVE(read_scan) {
 
   return ap_array;
 }
+
+PRIMITIVE(get_ap_info) {
```
Comment (kasperl) [inline src/resources/wifi_esp32.cc:593]:
```suggestion
PRIMITIVE(ap_info) {
```

File: system/extensions/esp32/wifi.toit [hunk 9]
```diff
@@ -247,8 +247,11 @@ class WifiModule implements NetworkModule:
     address_ = net.IpAddress ip
     logger_.info "network address statically assigned" --tags={"ip": address_}
 
-  rssi -> int?:
-    return wifi_get_rssi_ resource_group_
+  get_ap_info -> List:
+    result := []
+    ap_info := wifi_get_ap_info_ resource_group_
+    result.add_all ap_info
+    return result
```
Comment (kasperl) [inline system/extensions/esp32/wifi.toit:254]: The array returned from the primitive is a valid `List`.

```suggestion
  get_ap_info -> List:
    return wifi_get_ap_info_ resource_group_
```

File: system/extensions/esp32/wifi.toit [hunk 10]
```diff
@@ -45,9 +45,9 @@ class WifiServiceDefinition extends NetworkServiceDefinitionBase:
       return connect client arguments[0] arguments[1]
     if index == WifiService.ESTABLISH_INDEX:
       return establish client arguments
-    if index == WifiService.RSSI_INDEX:
+    if index == WifiService.GET_AP_INFO_INDEX:
```
Comment (kasperl) [inline system/extensions/esp32/wifi.toit:48]:
```suggestion
    if index == WifiService.AP_INFO_INDEX:
```

File: system/extensions/esp32/wifi.toit [hunk 11]
```diff
@@ -45,9 +45,9 @@ class WifiServiceDefinition extends NetworkServiceDefinitionBase:
...
       network := (resource client arguments) as NetworkResource
-      return rssi network
+      return get_ap_info network
```
Comment (kasperl) [inline system/extensions/esp32/wifi.toit:50]:
```suggestion
      return ap_info network
```

File: system/extensions/esp32/wifi.toit [hunk 12]
```diff
@@ -118,8 +118,8 @@ class WifiServiceDefinition extends NetworkServiceDefinitionBase:
   address resource/NetworkResource -> ByteArray:
     return (state_.module as WifiModule).address.to_byte_array
 
-  rssi resource/NetworkResource -> int?:
-    return (state_.module as WifiModule).rssi
+  get_ap_info resource/NetworkResource -> List:
+    return (state_.module as WifiModule).get_ap_info
```
Comment (kasperl) [inline system/extensions/esp32/wifi.toit:122]: Is this where the exception is thrown if we're not connected? Is it a nice and understandable exception?

```suggestion
  ap_info resource/NetworkResource -> List:
    return (state_.module as WifiModule).ap_info
```

**Review by kasperl — COMMENTED** [review]
Good stuff! I have a few suggestions:

File: lib/system/api/wifi.toit (hunk 4 above)
Comment (donghengqaz) [reply lib/system/api/wifi.toit:10]: Does it mean current we should not modify this version definition?

File: lib/net/wifi.toit (hunk 5 above)
Comment (donghengqaz) [reply lib/net/wifi.toit:149]: Getting value directly looks better.

File: src/resources/wifi_esp32.cc [hunk 13]
```diff
@@ -604,6 +589,35 @@ PRIMITIVE(read_scan) {
...
+
+  wifi_ap_record_t ap_record;
+  esp_err_t ret = esp_wifi_sta_get_ap_info(&ap_record);
+  if (ret != OK) return Primitive::os_error(ret, process);
+
+  const size_t element_count = 5;
+  Array* ap_array = process->object_heap()->allocate_array(element_count, Smi::zero());
+  if (ap_array == null) ALLOCATION_FAILED;
+
+  String* ssid = process->allocate_string((char *)ap_record.ssid);
```
Comment (kasperl) [inline src/resources/wifi_esp32.cc:604]:
```suggestion
  String* ssid = process->allocate_string((char*)ap_record.ssid);
```

File: src/resources/wifi_esp32.cc [hunk 14]
```diff
@@ -604,6 +589,35 @@ PRIMITIVE(read_scan) {
...
+  if (ssid == null) ALLOCATION_FAILED;
+
+  size_t bssid_size = 6;
```
Comment (kasperl) [inline src/resources/wifi_esp32.cc:607]: For consistency with `element_count`:

```suggestion
  const size_t bssid_size = 6;
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

**Pull request merged**